### PR TITLE
Add Playwright test suite (Fixes #14191)

### LINKED ISSUE #14191: Evaluate Selenium alternatives
## Description

We've used Selenium for our functional test runs for quite some time (tests are run via pytest-selenium). Generally it has been fairly reliable, but it does have some drawbacks that I think many of the team have felt:

1. Test written in Python are not always easy for front-end developers to write / update.
2. Test failures in CI are often hard to debug, with deep python stack traces to pick through.
3. Tests can be hard to instrument, with code often needing explicit waits added before making assertions to avoid flaky tests.
4. Updates can be painful. We've had to navigate many breaking changes over the years.

We've continued with Selenium mainly because it still supports IE testing via Sauce Labs. However IE is no longer supported today, and it becomes difficult to use newer versions of Selenium whilst falling back to older versions in CI. It may be time before long where we have to make a call to abandon automated IE testing altogether, which then opens up options to look at newer alternatives targeting only more modern browsers.

Given the above, I think it's a good time to look at some possible alternatives, as there are quite a few open source options now available:

- [Puppeteer](https://pptr.dev/)
- [Cypress](https://www.cypress.io/)
- [WebdriverIO](https://webdriver.io/)
- [Playwright](https://playwright.dev/)

The goal of this issue is to evaluate options, and then to pick one that best matches our needs. Ideally, it would also be great to get a test suite running against one of our deployments.

Testing criteria:

- Tests can run in either real browsers or headless in different browser engines.
- Tests can be written by front-end developers using JavaScript.
- Tests can target different browsers e.g. Firefox vs non-Firefox.
- Tests can be parameterized.
- Tests can run in CI against deployments (e.g. via GitHub actions)

---

#### Success Criteria
- [x] Evaluate alternatives to Selenium and pick one to test more closely.
- [x] Write a test suite which we can run against a deployment for evaluation purposes.

### PR DESCRIPTION
**Pull request by alexgibson**
## One-line summary

This PR adds a new suite of integration tests written using https://playwright.dev/. 

## Significant changes and points to review

Right now, this is only run manually and is not (yet) part of our CI deployment process. Once this merges, the next step is to start to run this against a deployment for a while to evaluate it for stability.

Aims of this new test suite:

1. It's written in JavaScript as opposed to Python, so will hopefully be more familiar for front-end developers to maintain.
2. It should be significantly easier to get running locally (everything is managed via an NPM install).
3. It hopefully adds more valuable tests (e.g. asserting auto file downloads, testing older UA strings) compared to the Selenium test suite (which had quite a number of old, superfluous tests for less critical pages).
4. Runs tests in all three major browser engines (Gecko, Chromium, Webkit).

If we like this better than Selenium and it's stable enough, we can look at potentially retiring the old test suite.

## Issue / Bugzilla link

https://github.com/mozilla/bedrock/issues/14191

## Testing

1. Start bedrock locally via `npm start`.
2. In a new terminal tab navigate to the playwright directory: `cd tests/playwright/`.
3. Install playwright via `npm install` (I opted for a separate `package.json` as the browser binaries are quite large).
5. Download latest version of browser binaries: `npm run install-deps`.

You should then be able to run Playwright tests (see the `./tests/playwright/README.md` file for instructions).

- [ ] Whole test suite should pass when running`npx playwright test`.

### CODE REVIEW

**Review by reemhamz — CHANGES_REQUESTED** [review]
This test suite was a small learning curve compared to Selenium (at least from a front-end perspective), and I appreciate how readable this test suite is! Good job researching for a test suite such as this and implementing it so well 👍🏼 

I read through each file individually and it all seems easier to grasp than what we currently have. 

Now, most of the tests pass, however the only page failing is the `/firefox/all` page, and I think it may have something to do with the [URL updating parameters based on the user's system ](https://github.com/mozilla/bedrock/blob/main/bedrock/firefox/tests/test_helpers.py#L503-L515) (i.e. if you're on desktop, `/#product-desktop-release/` gets added to the URL). 

The main error also showing up mentions that the test timeout is exceeded while waiting on the page to open:

<kbd>Screenshot 2024-06-18 at 3 46 15 PM</kbd>

**Comment by alexgibson** [comment]
@reemhamz thanks for taking this for a spin, I appreciate it! I'm glad that you could get things up and running ok with only limited instructions and the tests weren't too complicated to follow 😄 

The failing test you mentioned passes OK for me, so my guess here is that the timeout you encountered is because bedrock can get slowed down processing requests for this page, which is currently not very performant. Sometimes we see this same kind of failure in Selenium too in our live tests runs.

With that in mind though, this should improve once https://github.com/mozilla/bedrock/pull/14324 merges. For now, I'll skip these /all/ tests and will refactor once that PR merges 👍

**Comment by reemhamz** [comment]
Sounds good to me :) I'll approve the PR and I'll leave it to you to merge for when you think it's ready. Great job again!

**Review by reemhamz — APPROVED** [review]
I just noticed some test IDs missing in HTML files, however tests are passing so I don't think it's a blocking issue, but I'd love a bit of an explanation as to why some mof them are missing :O

File: tests/playwright/specs/products/firefox/firefox-developer.spec.js
Comment (alexgibson) [reply tests/playwright/specs/products/firefox/firefox-developer.spec.js:26]: The test IDs for download buttons are added automatically in: `bedrock/firefox/templates/firefox/includes/download-button.html`

image

File: tests/playwright/specs/products/firefox/firefox-installer-help.spec.js
Comment (alexgibson) [reply tests/playwright/specs/products/firefox/firefox-installer-help.spec.js:30]: See reply above :)

**Pull request merged**
